src/player.cpp

```diff
@@ -9039,82 +9039,59 @@ void player::use_fire(const int quantity)
     }
 }
 
-// does NOT return anything if the item is integrated toolset or fire!
-std::list<item> player::use_charges(itype_id it, long quantity)
+std::list<item> player::use_charges( const std::string& what, long qty )
 {
-    std::list<item> ret;
-    // the first two cases *probably* don't need to be tracked for now...
-    if (it == "toolset") {
-        charge_power(-quantity);
-        return ret;
-    }
-    if (it == "fire") {
-        use_fire(quantity);
-        return ret;
-    }
-    if ( it == "UPS" ) {
-        const long charges_off = std::min( quantity, charges_of( "UPS_off" ) );
-        if ( charges_off > 0 ) {
-            std::list<item> tmp = use_charges( "UPS_off", charges_off );
-            ret.splice(ret.end(), tmp);
-            quantity -= charges_off;
-            if (quantity <= 0) {
-                return ret;
-            }
+    std::list<item> res;
+
+    if( qty <= 0 ) {
+        return res;
+
+    } else if( what == "toolset" ) {
+        charge_power( -qty );
+        return res;
+
+    } else if( what == "fire" ) {
+        use_fire( qty );
+        return res;
+
+    } else if( what == "UPS" ) {
+        if( power_level > 0 && has_active_bionic( "bio_ups" ) ) {
+            auto bio = std::min( long( power_level ) , qty / 10 + ( qty % 10 != 0 ) );
+            charge_power( -bio );
+            qty -= std::min( qty, bio * 10 );
         }
-        return ret;
-    }
-    if (weapon.use_charges(it, quantity, ret)) {
-        remove_weapon();
-    }
-    for( auto a = worn.begin(); a != worn.end() && quantity > 0; ) {
-        if( a->use_charges( it, quantity, ret ) ) {
-            a = worn.erase( a );
-        } else {
-            ++a;
+
+        auto adv = charges_of( "adv_UPS_off", ceil( qty * 0.6 ) );
```

Comment:
`charges_of` may return values larger than `limit`. It does `qty += e->charges;` and later simply returns `qty`. If `limit` is 1, and `e->charges` is 10, it will return 10.

`adv` may yield more than `ceil( qty * 0.6 )` that way.